### PR TITLE
Thread SessionID through the handshake.

### PR DESCRIPTION
**Pull request by sayrer**
Takes `session_id` out of `HandshakeDetails`

### CODE REVIEW

File: rustls/src/client/hs.rs [hunk 1]
```diff
@@ -120,6 +120,7 @@ struct InitialState {
     dns_name: webpki::DNSName,
     transcript: HandshakeHash,
     extra_exts: Vec<ClientExtension>,
+    session_id: SessionID,
```
Comment (djc) [inline rustls/src/client/hs.rs:123]: I feel like this shouldn't exist in InitialState, and we should use Option instead of empty() and is_empty().

File: rustls/src/client/tls12.rs [hunk 2]
```diff
@@ -25,12 +25,14 @@ use crate::client::common::{ClientAuthDetails, ReceivedTicketDetails};
 use crate::client::common::{HandshakeDetails, ServerCertDetails, ServerKXDetails};
 use crate::client::hs;
 
+use crate::internal::msgs::handshake::SessionID;
```
Comment (djc) [inline rustls/src/client/tls12.rs:28]: This should be sorted with the other handshake type imports, and not go through the internal reexport.

**Review by djc — COMMENTED** [review]
Thanks for moving this forward!

File: rustls/src/client/tls12.rs (hunk 2 above)
Comment (sayrer) [reply rustls/src/client/tls12.rs:28]: Oh, sure. I think Intellij did this automatically. Oops.

File: rustls/src/client/hs.rs (hunk 1 above)
Comment (sayrer) [reply rustls/src/client/hs.rs:123]: That makes sense, I'll get this in the morning. (it's late in Pacific time)

Comment (briansmith) [reply rustls/src/client/hs.rs:123]: > I feel like this shouldn't exist in InitialState, and we should use Option instead of empty() and is_empty().

I agree that it seems like it shouldn't exist in `InitialState`.

I also agree that eventually we should use `Option<SessionID>` instead, but I don't think it has to be done in this PR.

Comment (djc) [reply rustls/src/client/hs.rs:123]: I suppose we could leave that until we can do the same transformation on the server side.

Comment (sayrer) [reply rustls/src/client/hs.rs]: Well, I did this before I saw these comments... lmk if the new diff is ok. It wasn't that hard to thread it through.

File: rustls/src/client/hs.rs [hunk 3]
```diff
@@ -341,12 +346,14 @@ fn emit_client_hello_for_retry(
         .map(ClientExtension::get_type)
         .collect();
 
+    let s_id = session_id.unwrap_or(SessionID::empty());
```
Comment (djc) [inline rustls/src/client/hs.rs:349]: Hmm, this defeats the purpose IMO. We should either go all the way and replace all uses of `empty()` and `is_empty()` with `None` and `is_none()` (in client) or keep it all as is.

(I would also prefer to avoid unrelated formatting changes.)

File: rustls/src/client/hs.rs (hunk 1 above)
Comment (djc) [reply rustls/src/client/hs.rs]: To remove those methods I guess we'd have to do the server side, too? But I was definitely angling to have the use of those methods decimated from the client side handshake code.

File: rustls/src/client/hs.rs (hunk 3 above)
Comment (sayrer) [reply rustls/src/client/hs.rs:349]: I don't think it does, since it gets the Option handling correct, up to this point. I'll look over the formatting changes.

File: rustls/src/client/hs.rs (hunk 1 above)
Comment (sayrer) [reply rustls/src/client/hs.rs]: I'll resolve this conversation and move discussion down below, to the current patch.

File: rustls/src/client/hs.rs (hunk 3 above)
Comment (djc) [reply rustls/src/client/hs.rs:348]: I guess that's true -- "defeats the purpose" is worded too strongly. Would you be up to change it throughout the rest of the client handshake?

Comment (sayrer) [reply rustls/src/client/hs.rs:348]: I think this is the best place to stop for one patch, but I am willing to keep going in another patch today. I'll have to change ClientHelloPayload to keep going, and I think that is best done separately.

Comment (djc) [reply rustls/src/client/hs.rs:348]: Fair enough.

Comment (djc) [reply rustls/src/client/hs.rs:348]: Can we just rename this `s_id` to `session_id` then? Suddenly giving it a different name doesn't seem helpful -- I think shadowing is more idiomatic in cases like this.

Comment (sayrer) [reply rustls/src/client/hs.rs]: Sure thing. Done.

**Comment by djc** [comment]
Thanks for sticking with it through all the back and forth!

File: rustls/src/client/hs.rs [hunk 4]
```diff
@@ -625,7 +632,7 @@ impl State for ExpectServerHello {
         server_hello
             .random
             .write_slice(&mut self.randoms.server);
-        self.handshake.session_id = server_hello.session_id;
+        self.session_id = server_hello.session_id;
```
Comment (briansmith) [inline rustls/src/client/hs.rs:635]: NIT: IMO this would have been better as `let session_id = `; there's no need to modify `self` in this function and the eventual goal is to make all these functions as close to purely functional as possible

Comment (djc) [reply rustls/src/client/hs.rs:635]: Fair enough; we'll pick it up in #596.
